schema_salad/tests/test_codegen_errors.py

```diff
@@ -0,0 +1,220 @@
+"""Tests of helpful error messages."""
+
+from pathlib import Path
+from typing import Any, MutableSequence, Optional, Union, cast
+from urllib.parse import urlparse
+
+import pytest
+
+import schema_salad.tests.cwl_v1_0 as cwl_v1_0
+from schema_salad.exceptions import ValidationException
+from schema_salad.utils import yaml_no_ts
+
+from .util import get_data
+
+
+def test_error_message1() -> None:
+    t = "test_schema/test1.cwl"
+    match = r"""^.*test1\.cwl:2:1:\s+Object\s+`.*test1\.cwl`\s+is\s+not\s+valid\s+because
+\s+tried\s+`Workflow`\s+but
+\s+\*\s+missing\s+required\s+field\s+`inputs`
+\s+\*\s+missing\s+required\s+field\s+`outputs`
+\s+\*\s+missing\s+required\s+field\s+`steps`"""
+    path = get_data("tests/" + t)
+    assert path
+    with pytest.raises(ValidationException, match=match):
+        load_document_by_uri(path)
+
+
+def test_error_message2() -> None:
+    t = "test_schema/test2.cwl"
+    match = r"""^.*test2\.cwl:2:1:\s+Field\s+`class`\s+contains\s+undefined\s+reference\s+to
+\s+`file://.+/schema_salad/tests/test_schema/xWorkflow`$"""
+    path = get_data("tests/" + t)
+    assert path
+    with pytest.raises(ValidationException, match=match):
+        load_document_by_uri(path)
+
+
+def test_error_message4() -> None:
+    t = "test_schema/test4.cwl"
+    match = r"""^.*test4.cwl:2:1:\s+Object\s+`.*test4.cwl`\s+is\s+not\s+valid\s+because
+\s+tried\s+`Workflow`\s+but
+.*test4\.cwl:6:1:\s+the\s+`outputs`\s+field\s+is\s+not\s+valid\s+because:
+.*test4\.cwl:7:3:\s+checking\s+object\s+`.*test4\.cwl#bar`
+\s+tried\s+`WorkflowOutputParameter`\s+but
+\s+the\s+`type`\s+field\s+is\s+not\s+valid\s+because:
+\s+Expected\s+one\s+of\s+\(list|dict|str,list|dict|str,list|dict|str\)\s+was
+\s+<class\s+'int'>"""
+    path = get_data("tests/" + t)
+    assert path
+    with pytest.raises(ValidationException, match=match):
+        load_document_by_uri(path)
+
+
+def test_error_message5() -> None:
```

Comment:
```
tests/test_schema/test5.cwl:2:1: Object `test5.cwl` is not valid because
                                   tried `Workflow` but
tests/test_schema/test5.cwl:8:1:     the `steps` field is not valid because:
                                       tried `array<WorkflowStep>` but
                                         - tried `array<WorkflowStep>` but
                                           Expected a list, was <class 'int'>
                                         - tried `WorkflowStep` but
                                           Expected a dict, was <class 'int'>
```

This error doesn't really make sense.  I don't know what `tried 'array<WorkflowStep>' but Expected a list, was <class 'int'>` is trying to tell me.

In this case the original validator is not great either:

```
test_schema/test5.cwl:2:1: Object 'test_schema/test5.cwl' is not valid because
                             tried 'Workflow' but
test_schema/test5.cwl:8:1:     the 'steps' field is not valid because
                                 tried array of <WorkflowStep> but
test_schema/test5.cwl:8:9:         item is invalid because
                                     is not a dict. Expected a WorkflowStep object.
```

Ideally it should be something like this:

```
tests/test_schema/test5.cwl:2:1: Object `test5.cwl` is not valid because
                                   tried `Workflow` but
tests/test_schema/test5.cwl:8:1:     the `steps` field is not valid because:
                                       tried `array<WorkflowStep>` but
                                         array item is invalid because
                                         expected a `WorkflowStep` object, but was <class 'int'>
```

Also, it would be nice if it just said `int` and not `<class 'int'>`.

Reply:
On that last point, I recommend a helper method that takes all the primitive objects that might show up in the YAML document (string, int, float, list, dict, None, True, False, CommentSeq, CommentedMap) and describes them consistently with the JSON/YAML/schema salad terminology:

string/int/float - same
list, CommentedSeq - array
dict, CommentedMap - object
True, False - boolean
None - null

Reply:
For the mapping, should that transform the error message for `test4.cwl` to 
```
test_schema/test4.cwl:2:1: Object `test4.cwl` is not valid because
test_schema/test4.cwl:6:1:   the `outputs` field is not valid because:
test_schema/test4.cwl:7:3:     checking object `test4.cwl#bar`
                                 the `type` field is not valid because:
                                   Expected one of (array, object, str) was int
```
dict -> object
list -> array

Reply:
This was the message that caused the most trouble for me. 


This first issue was in getting rid of the
```
- tried `array<WorkflowStep>` but
Expected a list, was <class 'int'>
```
message. 

In `ArrayLoader` objects of the array are loaded like,
```
lf = load_field(
                	doc[i], _UnionLoader(([self, self.items])), baseuri, loadingOptions, lc=lc
            	)
```
Which means an array of WorkflowStep will try loading the step with `array<WorkflowStep>` as well as `WorkflowStep`. This is what causes the messaging.

I first tried just having the loader be like

```
lf = load_field(
                	doc[i], self.items, baseuri, loadingOptions, lc=lc
            	)
```
but this caused an issue that certain fields weren't being passed through UnionLoader that were supposed to be, and "checking object {obj}" with lc info wasn't in error messages.

I then tried
```
lf = load_field(
                	doc[i], UnionLoader(self.items), baseuri, loadingOptions, lc=lc
            	)
```
which worked perfectly for all the test cases, but caused metaschema tests to fail.

Other similar changes (trying to suppress array error messages for example) cause downstream consequences

The approach I came up with that seemed to work with minimal changes, is in `UnionLoader`, if `t.alternates` includes something like `array<WorkflowStep>` as well as other items, suppress the messages coming from `array<WorkflowStep>`.

The other issue I had was with the generic object vs specific object type.

I was able to get error message 5 to read
```
schema_salad.exceptions.ValidationException: test_schema/test5.cwl:2:1: Object `test5.cwl` is not valid because:
test_schema/test5.cwl:8:1:   the `steps` field is not valid because:
test_schema/test5.cwl:8:9: 	array item is invalid because
                             	Expected a `WorkflowStep` object, was int
```

But that is at the expense of messages such as 4 being more verbose like
```
schema_salad.exceptions.ValidationException: test_schema/test4.cwl:2:1: Object `test4.cwl` is not valid because:
test_schema/test4.cwl:6:1:   the `outputs` field is not valid because:
                           	array item is invalid because
test_schema/test4.cwl:7:3:   	checking object `test4.cwl#bar`
                               	the `type` field is not valid because:
                                 	Expected one of (`OutputRecordSchema`, str,
                                 	`OutputArraySchema`, `OutputEnumSchema`) was int
```

versus

```
schema_salad.exceptions.ValidationException: test_schema/test5.cwl:2:1: Object `test5.cwl` is not valid because:
test_schema/test5.cwl:8:1:   the `steps` field is not valid because:
test_schema/test5.cwl:8:9: 	array item is invalid because
                             	Expected an object, was int
```
and
```
schema_salad.exceptions.ValidationException: test_schema/test4.cwl:2:1: Object `test4.cwl` is not valid because:
test_schema/test4.cwl:6:1:   the `outputs` field is not valid because:
                           	array item is invalid because
test_schema/test4.cwl:7:3:   	checking object `test4.cwl#bar`
                               	the `type` field is not valid because:
                                 	Expected one of (object, str) was int
```
I left it the later way for now (less verbose) but can change it over.